### PR TITLE
refactor: enable -Wconversion warning for screen.c

### LINKED ISSUE #567: Enable -Wconversion file by file
As illustrated here: https://github.com/tarruda/neovim/commit/86ea80dbd3b2a704c897e02d1d6ed8f01b92c96b#commitcomment-5984214 it often easy to make mistakes like assigning a variable of a wider type to a variable of a smaller type (`size_t` to `uint32_t` for example). The compiler doesn't warn about this without the `-Wconversion` flag. (in this case it wouldn't have warned even with the flag on on travis, which is 32-bits, but it would warn for anyone with a 64-bits OS, where `size_t` is usually 64-bits wide).

Sadly, the vim/neovim codebase has a lot (1000s of warnings) of things like this, and I'd wager that 99% of these assignments and over-casting are not really a problem (vim seems to be rather stable and works on a lot of platforms), but it's a class of issues that's easy to detect before they even take hold.

Since it's unfeasible for one person to do all this work on him/herself, it was proposed that this would be done on a file by file basis. With the workflow like this:
1. Create a branch
2. Enable `-Wconversion` for a file and fix the resulting warnings (1 commit)
3. Find another file that doesn't have `-Wconversion` enabled yet and goto 2 as long as you feel like it
4. Send PR

Also: any new files should be `-Wconversion` checked from the beginning.

Any takers just inform the rest of the team here!

@jszakmeister will hopefully be able to conjure up the groundwork for file-by-file enabling of certain flags once he's back from vacation :).

<bountysource-plugin>

---

Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/1740025-enable-wconversion-file-by-file?utm_campaign=plugin&utm_content=tracker%2F461131&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F461131&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by dundargoc**
Closes https://github.com/neovim/neovim/issues/567

### CODE REVIEW

File: src/nvim/screen.c [hunk 1]
```diff
@@ -2734,7 +2734,7 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
                 for (char_u *p1 = (char_u *)skipwhite((char *)extra); p1 < p2; p1++, p2--) {
                   const int t = *p1;
```
Comment (zeertzjq) [inline src/nvim/screen.c:2735]:
```suggestion
                  const char_u t = *p1;
```

File: src/nvim/screen.c [hunk 2]
```diff
@@ -2734,7 +2734,7 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
...
                   *p1 = *p2;
-                  *p2 = t;
+                  *p2 = (char_u)t;
```
Comment (zeertzjq) [inline src/nvim/screen.c:2737]:
```suggestion
                  *p2 = t;
```

File: src/nvim/screen.c [hunk 3]
```diff
@@ -3838,7 +3838,7 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
         if (n != 0) {
           /* At the window boundary, highlight the last character
            * instead (better than nothing). */
-          off += n;
+          off += (unsigned)n;
```
Comment (zeertzjq) [inline src/nvim/screen.c:3841]: This can cast a negative value to `unsigned`

File: src/nvim/screen.c [hunk 4]
```diff
@@ -3973,7 +3973,7 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
         while (col >= 0 && col < grid->cols) {
           schar_from_ascii(linebuf_char[off], ' ');
           linebuf_attr[off] = vcol >= TERM_ATTRS_MAX ? 0 : term_attrs[vcol];
-          off += n;
+          off += (unsigned)n;
```
Comment (zeertzjq) [inline src/nvim/screen.c:3976]: This can cast a negative value of `unsigned`

File: src/nvim/screen.c [hunk 5]
```diff
@@ -4536,7 +4536,7 @@ void rl_mirror(char_u *str)
   for (p1 = str, p2 = str + STRLEN(str) - 1; p1 < p2; ++p1, --p2) {
     t = *p1;
     *p1 = *p2;
-    *p2 = t;
+    *p2 = (char_u)t;
```
Comment (zeertzjq) [inline src/nvim/screen.c:4539]: It's probably better to change the type of `t` to `char_u`

File: src/nvim/screen.c (hunk 2 above)
Comment (zeertzjq) [inline src/nvim/screen.c:2737]: It's probably better to change the type of `t` to `char_u`

File: src/nvim/screen.c (hunk 3 above)
Comment (dundargoc) [reply src/nvim/screen.c:3841]: changing `off` to `int` then.

**Pull request merged**